docs/design/settings/extensions/modular-settings-extensions.md

```diff
@@ -0,0 +1,314 @@
+]
+default-version = "v2"
+```
+
+Any setting can be written or read at any of the `supported-versions` listed by the extension.
+Unversioned requests to read or write this setting are assumed to be the `default-version`.
+During a `set` transaction, the `apiserver` will consider the provided version to be the `canonical-version`, and will ask the settings extension to perform migrations from the `canonical` version to all supported versions using the `migrate` command.
+These migrations always begin at the `canonical` version, and settings extensions are free to perform the migrations however they wish; however, Bottlerocket's settings SDK will provide utilities to help settings developers support a series of linear setting upgrades and downgrades.
+
+All migrations are performed and stored at setting-write-time because Bottlerocket rollbacks may result in running an OS running with a settings extension which does not support the version at which a given setting is stored.
+If Bottlerocket boots and finds that a setting is stored at a value which is not supported by it's installed setting extension, it will allow that version to persist unless a new value is `set` by the `apiserver`, at which point it is removed -- this is to attempt to prevent dataloss in the case that the image can eventually update to a new image which supports the stored setting version.
+
+### Extensions and Templating
+Fetching settings from the Bottlerocket API/datastore is currently done via two mechanisms:
+* Fetching data directly from the API using `apiclient`
+* Rendering config files (or just strings) using `schnauzer`
+
+For open-source tools, Bottlerocket prefers the rendered config file approach, as it doesn't require patching upstream code in order to consume the Bottlerocket API.
+First-party tools have no such constraints, and so there exists tools that take either approach.
+
+Because settings are discovered at runtime on the system, and indeed may no longer be handled by models written in Rust, it will no longer be feasible to create an `apiclient` with the same strong typing as exists today.
+It will instead be preferred that system services, including first-party services, all route their fetching of API settings through their own configuration files, which will be rendered by Bottlerocket whenever the related settings are changed.
+
+Today, configuration templates are rendered by `schnauzer` using [handlebars](https://handlebarsjs.com/) templates.
+`schnauzer` fetches the entirety of the settings tree, defines some useful helper functions, and then uses these as context for rendering the template.
+With the introduction of settings extensions, this will change such that templates specify `(name, version)` tuples for all extensions needed to render the template.
+These extension specifications will be given within the template as "front matter" -- a TOML block specified at the head of the document and delineated with `---` as such:
+
+```toml
+---
```

Comment:
I would suggest delimiting front mater with something other than triple dash.

Triple plus (`+++`) is my preference as [Hugo, Forestry, and Zola use it. Some other software uses `---toml`](https://github.com/denoland/deno_std/issues/3094#issuecomment-1400018219) but that is not symmetrical, the ending being triple dash.. yuck. 

Triple dash is likely to confuse people as that alone is widely used for yaml.

Reply:
I would prefer not to use frontmatter since I'm concerned it will confuse syntax highlighters, linters, etc. Case in point, see the red highlights in this GitHub render. It rather see a separate file with a naming convention, theconfig.meta.toml or theconfig.frontmatter.toml... something like that.